fgpyo/io/io.py

```diff
@@ -0,0 +1,175 @@
+   >>> IO.path_is_readable(path_flat)
+   AssertionError: Cannot read non-existent path: example.txt
+   >>> IO.path_is_readable(compressed_file)
+   AssertionError: Cannot read non-existent path: example.txt.gz
+   Write to and read from path
+   >>> write_lines(path = path_flat, lines_to_write=["flat file", 10])
+   >>> write_lines(path = path_compressed, lines_to_write=["gzip file", 10])
+   Read lines from a path into a generator
+   >>> read_lines(path = path_flat)
+   ['flat file', '10']
+   >>> read_lines(path = path_compressed)
+   ['gzip file', '10']
+"""
+
+import gzip
+import io
+import os
+from pathlib import Path
+from typing import IO
+from typing import Any
+from typing import Generator
+from typing import Iterable
+from typing import Iterator
+from typing import Set
+from typing import TextIO
+from typing import Union
+from typing import cast
+
+COMPRESSED_FILE_EXTENSIONS: Set[str] = {".gz", ".bgz"}
+
+
+def path_is_readable(path: Path) -> None:
+    """Checks that file exists and returns True, else raises FileNotFoundError
+
+    Args:
+        path: a Path to check
+
+    Example:
+    _file_exists(path = Path("some_file.csv"))
+    """
+    assert path.exists(), f"Cannot read non-existent path: {path}"
+    assert not path.is_dir(), f"Cannot read path becasue it is a directory: {path}"
+    assert os.access(path, os.R_OK), f"Path exists but is not readable: {path}"
+
+
+def directory_exists(path: Path) -> None:
+    """Asserts that a path exist and is a directory
+
+    Args:
+        path: Path to check
+
+    Example:
+    _directory_exists(path = Path("/example/directory/"))
+    """
+    assert path.exists(), f"Path does not exist: {path}"
+    assert path.is_dir(), f"Path exists but is not a directory: {path}"
+
+
+def path_is_writeable(path: Path, parent_must_exist: bool = True) -> None:
+    """Checks that path is writable and returns True, else raises an Exception
+    Args:
+         path: Path to check
+         parent_must_exist: True/False the parent directory must exist, the default is True
+    Example:
+    _writeable(path = Path("example.txt"))
+    """
+    assert path.exists(), f"Cannot write file because path is non-existent: {path}"
+    assert path.is_file(), f"Cannot write file because path is a directory: {path}"
+    assert os.access(path, os.W_OK), f"File exists but is not writebale: {path}"
+
+    if not parent_must_exist:
+        return None
+
+    parent: str = f"{path.parent.absolute()}"
+    assert Path(
+        parent
+    ).exists(), f"Cannot write file because parent diretory does not exist: {path}"
+    assert Path(
+        parent
+    ).is_dir(), f"Cannot write file because parent exists and is not a directory: {path}"
+    assert os.access(
+        parent, os.W_OK
+    ), f"Cannot write file because parent directory is not writeable: {path}"
+
+
+def to_reader(path: Path) -> Union[io.TextIOWrapper, TextIO, IO[Any]]:
+    """Opens a Path for reading with a specified mode or default 'rt'.
```

Comment:
I don't see a `mode` parameter that this implies.  This should either take a mode parameter (which is open to abuse - what if a user passes `w` as the mode?) or should be called `def open_for_text_reading(path: Path)`.